wdl-lint/src/rules/runtime_section_keys.rs

```diff
@@ -0,0 +1,502 @@
+        _: &TaskDefinition,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.current_runtime_span = None;
+                self.encountered_keys = Default::default();
+            }
+            VisitReason::Exit => {
+                // If a runtime section span has not been encountered, then
+                // presumably there won't be any place to put missing
+                // recommended runtime keys.
+                let runtime_span = match self.current_runtime_span {
+                    Some(span) => span,
+                    None => return,
+                };
+
+                // SAFETY: at this point in the execution rule, the document
+                // version should have been encountered and set (for the outer
+                // [`Option`]).
+                if let Some(SupportedVersion::V1(minor_version)) =
+                    self.version.expect("the document version should be known")
+                {
+                    let specification = format!("the {minor_version} specification");
+
+                    if !self.non_reserved_keys.is_empty() {
+                        state.add(report_non_reserved_runtime_keys(
+                            &self.non_reserved_keys,
+                            runtime_span,
+                            &specification,
+                        ));
+                    }
+
+                    let recommended_keys = match minor_version {
+                        V1::Zero => recommended_keys(keys_v1_0()),
+                        V1::One => recommended_keys(keys_v1_1()),
+                        _ => return,
+                    };
+
+                    let missing_keys = recommended_keys
+                        .filter(|(key, _)| {
+                            !self.encountered_keys.iter().any(|s| s.as_str() == *key)
+                        })
+                        .map(|(key, _)| key)
+                        .collect::<Vec<_>>();
+
+                    if !missing_keys.is_empty() {
+                        state.add(report_missing_recommended_keys(
+                            missing_keys,
+                            runtime_span,
+                            &specification,
+                        ));
+                    }
+                }
+            }
+        }
+    }
+
+    fn runtime_section(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        section: &RuntimeSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        match self.current_runtime_span {
+            // Since validation and linting occur simultaneously, it is possible
+            // that we would hit this condition for a task with two `runtime`
+            // sections. Here, we simply ignore any `runtime` section but the
+            // first one, as multiple `runtime` sections will be flagged as an
+            // error by the validation.
+            Some(_) => {}
```

Comment:
So I think this isn't quite correct, as we'll still process the items of the additional runtime sections.

I think we should have a flag, reset on task entry, which indicates that we've exited from a runtime section; if that flag is set, have future calls to `runtime_section` and `runtime_item` return early.

Reply:
Good catch here, I corrected this.